includes/class-wc-payments-customer-service.php

```diff
@@ -47,12 +70,24 @@ public function get_customer_id_by_user_id( $user_id ) {
 			return null;
 		}
 
-		$customer_id = get_user_option( self::WCPAY_CUSTOMER_ID_OPTION, $user_id );
-		if ( false === $customer_id ) {
-			return null;
+		$customer_id = get_user_option( $this->get_customer_id_option(), $user_id );
+
+		// If customer_id is false it could mean that it hasn't been migrated from the deprecated key.
+		// If the gateway is in live mode, proceed to migrate the legacy customer ID to the new meta.
+		if ( false === $customer_id && ! WC_Payments::get_gateway()->is_in_test_mode() ) {
```

Comment:
To me `! WC_Payments::get_gateway()->is_in_test_mode()` looks like un-needed: we want to migrate all accounts, so the test systems are also not getting broke.

But we can do something like this:
```php
		if ( false === $customer_id ) {
			$customer_id = get_user_option( self::DEPRECATED_WCPAY_CUSTOMER_ID_OPTION, $user_id );
			if ( false !== $customer_id ) {
				$key_to_update = WC_Payments::get_gateway()->is_in_test_mode()
					? self::WCPAY_TEST_CUSTOMER_ID_OPTION
					: self::WCPAY_LIVE_CUSTOMER_ID_OPTION;
				if ( update_user_option( $user_id, $key_to_update, $customer_id ) ) {
					// A customer migration succeeded, hence dropping legacy meta
					delete_user_option( $user_id, self::DEPRECATED_WCPAY_CUSTOMER_ID_OPTION );
				} else {
					// A customer migration failed, preserve legacy meta
					Logger::error( sprintf( 'Failed to migrate customer ID for user %s, will retry next time.', $user_id ) );
				}
			}
		}
```

Reply:
I'm sorry to have not included some context in the PR description. Unfortunately, there's no easy way to know if a customer is test or live, as the current store mode cannot give us insights into the customer mode when it was created. We've discussed this in paJDYF-QQ-p2 and came to the conclusion that we'd migrate all customers to live mode and let them be deleted if they happen to be test customers, to avoid losing live customer data.

It's possible to fix an incorrect migration, but the code is brittle and we need to watch multiple entry-points, which is why https://github.com/Automattic/woocommerce-payments/pull/918#issuecomment-702847190 was closed. @dwainm raised the same concern in #973, so if we feel like we don't want to let test customers be deleted we could resuscitate #918 and take on the additional maintenance that it requires.

Reply:
Okay, then 2 points here:

- then we could replace `update_user_option( $user_id, $this->get_customer_id_option(), $customer_id )` with `update_user_option( $user_id, self::WCPAY_LIVE_CUSTOMER_ID_OPTION, $customer_id )` for clearing our intentions here.

- if I go to WCPay dev plugin setting, in "Account cache contents" I can see account details and especially the following:
```php
array (
  ...
  'is_live' => false,
  ...
)
```
Can we use the `is_live` field value to identify if it a live account?

Reply:
> then we could replace update_user_option( $user_id, $this->get_customer_id_option(), $customer_id ) with update_user_option( $user_id, self::WCPAY_LIVE_CUSTOMER_ID_OPTION, $customer_id ) for clearing our intentions here.

Sure! Done in https://github.com/Automattic/woocommerce-payments/pull/972/commits/0fed69f350f051c01db241e33ce384a41c16cec0.

> Can we use the is_live field value to identify if it a live account?

I think so. I hadn't thought about it, but we could use that value to salvage some test data from accounts that could never have a `live` object and default to `live` mode for the rest of them. Is that what you're suggesting?

Reply:
Thanks!

Yes. To reassure: if the account has `is_live` value, take advantage of it. Otherwise assume it's live (perfectly aligned with previous discussions).

Reply:
> Yes. To reassure: if the account has is_live value, take advantage of it. Otherwise assume it's live (perfectly aligned with previous discussions).

Awesome! I like that a lot, thanks for suggesting it! I'll take a stab at it and update this discussion with the commit once it's working.

Reply:
I've implemented in 8259954, 1e197f7 and 27cd3ae. What do you think?

Reply:
Awesome changes, really. Go ahead with updating tests + I left two remarks and I'll make a final manual test on Monday.